File: 3.Behavioral/3.5.Mediator/Mediator.Main.cpp
#include <iostream>
#include <string>

class Colleague;
class Mediator;
class ConcreteMediator;
class ConcreteColleague1;
class ConcreteColleague2;

class Mediator
{
public:
	virtual void Send(std::string const& message, Colleague *colleague) const = 0;
};

class Colleague
{
protected:
	Mediator* mediator_;

public:
	explicit Colleague(Mediator *mediator) :mediator_(mediator)
	{
	}

	virtual void Send(std::string const& message) = 0;
	virtual void Notify(std::string const& message) = 0;
};

class ConcreteColleague1 :public Colleague
{
public:
	explicit ConcreteColleague1(Mediator* mediator) 
		: Colleague(mediator)
	{
	}

	void Send(std::string const& message) override
	{
		mediator_->Send(message, this);
	}

	void Notify(std::string const& message) override
	{
		std::cout << "Colleague1 gets message '" << message << "'" << std::endl;
	}
};

class ConcreteColleague2 :public Colleague
{
public:
	explicit ConcreteColleague2(Mediator *mediator) :Colleague(mediator)
	{
	}

	void Send(std::string const& message) override
	{
		mediator_->Send(message, this);
	}

	void Notify(std::string const& message) override
	{
		std::cout << "Colleague2 gets message '" << message << "'" << std::endl;
	}
};

class ConcreteMediator :public Mediator
{
protected:
	Colleague      *m_Colleague1;
	Colleague      *m_Colleague2;

public:
	void SetColleague(Colleague *c)
	{
		if (!m_Colleague1)
			m_Colleague1 = c;
		else if (!m_Colleague2)
			m_Colleague2 = c;
	}

	virtual void Send(std::string const& message, Colleague *colleague) const override
	{
		if (colleague == m_Colleague1)
		{
			m_Colleague2->Notify(message);
		}
		else if (colleague == m_Colleague2)
		{
			m_Colleague1->Notify(message);
		}
	}
};

int main()
{
	ConcreteMediator  mediator;

	ConcreteColleague1 colleague_1(&mediator);
	ConcreteColleague2 colleague_2(&mediator);

	mediator.SetColleague(&colleague_1);
	mediator.SetColleague(&colleague_2);

	colleague_1.Send("How are you?");
	colleague_2.Send("Fine, thanks");

	std::cin.get();
	return 0;
}

/*
Output
Colleague2 gets message 'How are you?'
Colleague1 gets message 'Fine, thanks'
*/

File: 3.Behavioral/3.5.Mediator.cs/Mediator.Main.cs
// Mediator pattern — Structural example
using System;

namespace DoFactory.GangOfFour.Mediator.Structural
{
    /// <summary>
    /// MainApp startup class for Structural
    /// Mediator Design Pattern.
    /// </summary>
    class MainApp
    {
        /// <summary>
        /// Entry point into console application.
        /// </summary>
        static void Main()
        {
            ConcreteMediator mediator = new ConcreteMediator();

            ConcreteColleague1 collegue_1 = new ConcreteColleague1(mediator);
            ConcreteColleague2 collegue_2 = new ConcreteColleague2(mediator);

            mediator.collegue_1 = collegue_1;
            mediator.collegue_2 = collegue_2;

            collegue_1.Send("How are you?");
            collegue_2.Send("Fine, thanks");

            // Wait for user
            Console.ReadKey();
        }
    }

    /// <summary>
    /// The 'Mediator' abstract class
    /// </summary>
    abstract class Mediator
    {
        public abstract void Send(string message, Colleague colleague);
    }

    /// <summary>
    /// The 'ConcreteMediator' class
    /// </summary>
    class ConcreteMediator : Mediator
    {
        public ConcreteColleague1 collegue_1 { private get; set; }

        public ConcreteColleague2 collegue_2 { private get; set; }

        public override void Send(string message, Colleague colleague)
        {
            if (colleague == collegue_1)
            {
                collegue_2.Notify(message);
            }
            else
            {
                collegue_1.Notify(message);
            }
        }
    }

    /// <summary>
    /// The 'Colleague' abstract class
    /// </summary>
    abstract class Colleague
    {
        protected Mediator mediator;

        // Constructor
        public Colleague(Mediator mediator)
        {
            this.mediator = mediator;
        }
    }

    /// <summary>
    /// A 'ConcreteColleague' class
    /// </summary>
    class ConcreteColleague1 : Colleague
    {
        // Constructor
        public ConcreteColleague1(Mediator mediator)
          : base(mediator)
        {
        }

        public void Send(string message)
        {
            mediator.Send(message, this);
        }

        public void Notify(string message)
        {
            Console.WriteLine("Colleague1 gets message: " + message);
        }
    }

    /// <summary>
    /// A 'ConcreteColleague' class
    /// </summary>
    class ConcreteColleague2 : Colleague
    {
        // Constructor
        public ConcreteColleague2(Mediator mediator)
          : base(mediator)
        {
        }

        public void Send(string message)
        {
            mediator.Send(message, this);
        }

        public void Notify(string message)
        {
            Console.WriteLine("Colleague2 gets message: " + message);
        }
    }
}
/**
* Output
* Colleague2 gets message: How are you?
* Colleague1 gets message: Fine, thanks
*/